### PR TITLE
Anonymise ID column dialog

### LINKED ISSUE #4715: Hash function to anonymise ID column
I have written a function to anonymise an ID column by using a hash function from the `digest` package:

```
hashed_id <- function(x, salt, algo = "crc32") {
  y <- paste(x, salt)
  y <- sapply(y, function(X) digest::digest(X, algo = algo))
  as.character(y)
}
```
The salt argument is an arbitrary string which just increases the security as this should only be known by the user.

The usefulness of this over randomising the values or replacing them by 1 to n is that these are complex algorithms but still being reproducible and reversible with knowledge

Not urgent but this could be useful dialog under Prepare to Share and easy to create. This is not something that the `sdcMicro` package does and could be done before/after using sdcMicro and complements what it does.

### PR DESCRIPTION
**Pull request by shadrackkibet**
Fixes #4715

### CODE REVIEW

**Comment by shadrackkibet** [comment]
@africanmathsinitiative/developers this is ready for review and testing.

**Comment by shadrackkibet** [comment]
@dannyparsons  please have a look at this.

**Comment by shadrackkibet** [comment]
> There should also be an option for the algorithm to use.

A textbox?

**Comment by dannyparsons** [comment]
combobox, if you look at the digest function you'll see the options

**Comment by shadrackkibet** [comment]
@africanmathsinitiative/developers Can someone now test this?

**Comment by Patowhiz** [comment]
Testing this

**Comment by Patowhiz** [comment]
@shadrackkibet is this dialog on the frmMain menu? Where?

**Comment by shadrackkibet** [comment]
This is not yet added. You can add it somewhere and test it.

**Comment by Patowhiz** [comment]
The only other minor issue observed is the tab index. It skips the algorithm combobox and goes to it after the  to script button. Other than that this dialog functions well.

**Comment by dannyparsons** [comment]
Good to check the algorithm is being passed through correctly.

**Comment by shadrackkibet** [comment]
> @shadrackkibet I just noticed that with the same inputs , the crc32 and md5 algorithms give the same result. Is this intentional? I thought the internal implementation of these algorithms is different, I expected different results even with the same inputs.

This is not intentional. I am going to look into it.

**Comment by shadrackkibet** [comment]
@Patowhiz above bug has been sorted out now.

**Comment by shadrackkibet** [comment]
@Patowhiz is this now ready for @dannyparsons ?

**Comment by Patowhiz** [comment]
This is now fixed

**Comment by shadrackkibet** [comment]
@dannyparsons this is ready for you to review.

File: instat/dlgAnonymiseIDColumn.vb [hunk 1]
```diff
@@ -1,3 +1,100 @@
...
+    End Sub
+
+    Private Sub ucrBase_ClickReset(sender As Object, e As EventArgs) Handles ucrBase.ClickReset
+        SetDefaults()
+        SetRCodeForControls(True)
+        TestOKEnabled()
+    End Sub
+
+    Private Sub TestOKEnabled()
+        If Not ucrReceiverIDColumn.IsEmpty AndAlso Not ucrInputTextSalt.IsEmpty AndAlso ucrSaveAnonymisedColumn.IsComplete Then
```
Comment (dannyparsons) [inline instat/dlgAnonymiseIDColumn.vb]: Although I would be suggesting that salt is added, it isn't necessary to do this, so salt could be nothing.

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 2]
```diff
@@ -1084,4 +1084,10 @@ compare_columns <- function(x, y, use_unique = TRUE, sort_values = TRUE, firstno
     }
     if(display_union) cat(paste0("Union (Values that appear in either column): ", paste0("'", dplyr::union(x, y), "'", collapse = ", ")))
   }
+}
+
+hashed_id <- function(x, salt, algo = "crc32") {
+  y <- paste(x, salt)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/stand_alone_functions.R]: This also needs to change if salt is optional. The paste should only be done if salt is not missing.

Comment (shadrackkibet) [reply instat/static/InstatObject/R/stand_alone_functions.R]: Great, i am going to modify this.

**Comment by shadrackkibet** [comment]
@dannyparsons that is now done.

**Comment by shadrackkibet** [comment]
Where should this be on the menu?

**Comment by dannyparsons** [comment]
> Not urgent but this could be useful dialog under Prepare to Share https://github.com/africanmathsinitiative/R-Instat/issues/4715#issue-334512093

**Pull request merged**